rts/idris_rts.c

```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
+
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    unsigned int j;
+    for (j = 0; j < sizeof(int) * 4 + 1; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j])) && temp[j] != '-') {
+            fprintf(stderr, "Fatal error: invalid output from sprintf on an int!\n");
+            fprintf(stderr, "It was [%c] at index [%u] in [%s].\n", temp[j], j, temp);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return MKSTR(vm, (const utf8_byte*)temp);
 }
 
 VAL idris_castBitsStr(VM* vm, VAL i) {
-    Closure* cl;
     ClosureType ty = i->ty;
+    // max length 64 bit unsigned int str 20 chars (18,446,744,073,709,551,615)
+    char temp[21] = { 0 };
 
     switch (ty) {
     case BITS8:
-        // max length 8 bit unsigned int str 3 chars (256)
-        cl = allocate(sizeof(Closure) + sizeof(char)*4, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu8, (uint8_t)i->info.bits8);
+        sprintf(temp, "%" PRIu8, (uint8_t)i->info.bits8);
         break;
     case BITS16:
-        // max length 16 bit unsigned int str 5 chars (65,535)
-        cl = allocate(sizeof(Closure) + sizeof(char)*6, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu16, (uint16_t)i->info.bits16);
+        sprintf(temp, "%" PRIu16, (uint16_t)i->info.bits16);
         break;
     case BITS32:
-        // max length 32 bit unsigned int str 10 chars (4,294,967,295)
-        cl = allocate(sizeof(Closure) + sizeof(char)*11, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu32, (uint32_t)i->info.bits32);
+        sprintf(temp, "%" PRIu32, (uint32_t)i->info.bits32);
         break;
     case BITS64:
-        // max length 64 bit unsigned int str 20 chars (18,446,744,073,709,551,615)
-        cl = allocate(sizeof(Closure) + sizeof(char)*21, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu64, (uint64_t)i->info.bits64);
+        sprintf(temp, "%" PRIu64, (uint64_t)i->info.bits64);
         break;
     default:
         fprintf(stderr, "Fatal Error: ClosureType %d, not an integer type", ty);
         exit(EXIT_FAILURE);
     }
 
-    SETTY(cl, STRING);
-    return cl;
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < 21; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j]))) {
+            fprintf(stderr, "Fatal error: invalid output from sprintf on bit type %d!", ty);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return MKSTR(vm, (const utf8_byte *)temp);
 }
 
 VAL idris_castStrInt(VM* vm, VAL i) {
     char *end;
-    i_int v = strtol(GETSTR(i), &end, 10);
-    if (*end == '\0' || *end == '\n' || *end == '\r')
-        return MKINT(v);
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
```

Comment:
it is, but the size field is the size of the actual data, not including the `'\0'` at the end for C